src/externs.ts

```diff
@@ -160,18 +160,42 @@ export function generateExterns(
   const mtt =
       new ModuleTypeTranslator(sourceFile, typeChecker, host, diagnostics, /*isForExterns*/ true);
 
-  let rootNamespace = '';
+  // .d.ts files declare symbols. The code below translates these into a form understood by Closure
+  // Compiler, converting the type syntax, but also converting symbol names into a form accessible
+  // to Closure Compiler.
+
+  // Like regular .ts files, .d.ts can be either scripts or modules. Scripts declare symbols in the
+  // global namespace, which has the same semantics in Closure and TypeScript, so the code below
+  // emits those with the same name.
+
+  // Modules however declare symbols scoped to the module that can be exported. Closure has no
+  // concept of externs that are non-global, so tsickle needs to mangle the symbol names, both at
+  // their declaration and at their use site.
+
+  // This mangling happens by wrapping all declared symbols in a namespace based on the file name.
+  // This namespace is then essentially the exports object for the ambient module (externs in
+  // Closure terms). This namespace is called `moduleNamespace` below:
+  let moduleNamespace = '';
   if (isExternalModule) {
-    // .d.ts files that are modules do not declare global symbols - their symbols must be explicitly
-    // imported to be used. However Closure Compiler has no concept of externs that are modules and
-    // require imports. This code mangles the symbol names by wrapping them in a top level variable
-    // that's unique to this file. That allows emitting them for Closure as global symbols while
-    // avoiding collisions. This is necessary as symbols local to this module can (and will very
-    // commonly) conflict with the namespace used in "export as namespace", e.g. "angular", and also
-    // to avoid users accidentally using these symbols in .js files (and more collisions). The
-    // symbols that are "hidden" like that can be made accessible through an "export as namespace"
-    // declaration (see below).
-    rootNamespace = moduleNameAsIdentifier(host, sourceFile.fileName);
+    moduleNamespace = moduleNameAsIdentifier(host, sourceFile.fileName);
+  }
+
+  // Symbols are generated starting in rootNamespace. For script .d.ts with global symbols, this is
+  // the empty string. For most module `.d.ts` files, this is the mangled namespace object. The
+  // remaining special case are `.d.ts` files containing an `export = something;` statement. In
+  // these, the effective exports object, i.e. the object containing the symbols that importing code
+  // receives, is different from the main module scope.
+  // tsickle handles the `export =` case by generating symbols in a different namespace (escaped
+  // with a `_`) below, and then assigning whatever is actually exported into the `moduleNamespace`
+  // below.
```

Comment:
Testing my understanding of this code, I think that  `exports {foo: bar}` will also need some special handling. Add a TODO if I am right (not worth fixing now).